coap-endpoint/src/main.rs

```diff
@@ -0,0 +1,418 @@
+    Ok((path_segments, queries, auth))
+}
+
+async fn publish_handler<S>(
+    mut request: CoapRequest<SocketAddr>,
+    app: App<S>,
+) -> Option<CoapResponse>
+where
+    S: DownstreamSink + Send,
+    <S as DownstreamSink>::Error: Send,
+{
+    let path_segments: Vec<String>;
+    let queries: Option<&Vec<u8>>;
+    let auth: &Vec<u8>;
+
+    if let Ok((p, q, a)) = params(&request) {
+        path_segments = p;
+        queries = q;
+        auth = a;
+    } else {
+        let ret = Err(CoapEndpointError(EndpointError::InvalidRequest {
+            details: "Invalid Path".to_string(),
+        }))
+        .respond_to(&mut request);
+        return ret;
+    }
+
+    let options = queries
+        .map(|x| serde_urlencoded::from_bytes::<PublishOptions>(x).ok())
+        .flatten()
+        .unwrap_or_default();
+
+    match path_segments.len() {
+        1 => telemetry::publish_plain(
+            app.downstream,
+            app.authenticator,
+            app.commands,
+            path_segments[0].clone(),
+            options,
+            request.clone(),
+            auth,
+        )
+        .await
+        .respond_to(&mut request),
+
+        2 => telemetry::publish_tail(
+            app.downstream,
+            app.authenticator,
+            app.commands,
+            (path_segments[0].clone(), path_segments[1].clone()),
+            queries
+                .map(|x| serde_urlencoded::from_bytes::<PublishOptions>(x))?
+                .map_err(anyhow::Error::from)
+                .ok()?,
+            request.clone(),
+            auth,
+        )
+        .await
+        .respond_to(&mut request),
+
+        _ => Err(CoapEndpointError(EndpointError::InvalidRequest {
+            details: "Invalid Path".to_string(),
+        }))
+        .respond_to(&mut request),
+    }
+}
+
+// Health server uses actix_web to run
+#[actix_web::main]
+async fn main() -> anyhow::Result<()> {
+    env_logger::init();
+    dotenv().ok();
+
+    let config = Config::from_env()?;
+    let commands = Commands::new();
+    let addr = config.bind_addr_coap.unwrap_or("0.0.0.0:5683".to_string());
+    let coap_server_commands = commands.clone();
+
+    let app = App {
+        downstream: DownstreamSender::new(KafkaSink::new("DOWNSTREAM_KAFKA_SINK")?)?,
+        authenticator: DeviceAuthenticator(
+            drogue_cloud_endpoint_common::auth::DeviceAuthenticator::new().await?,
+        ),
+        commands: coap_server_commands,
+    };
+
+    println!("Server up on {}", addr);
+    let mut server = Server::new(addr).unwrap();
+
+    let device_to_endpoint = server.run(move |request| publish_handler(request, app.clone()));
+
+    let health = HealthServer::new(config.health, vec![]);
+
+    let mut command_server = CommandServer::new(config.command, commands)?;
+
+    futures::try_join!(
+        health.run(),
+        device_to_endpoint.err_into(),
+        command_server.deref_mut().err_into(),
+    )?;
+    Ok(())
+}
+
+#[cfg(test)]
+pub mod test {
```

Comment:
Maybe add a test here for parsing url-encoded device ids

Reply:
how exactly are devices+apps passed through the url? As params? The auth option number format doesn't really enforce any encoding, just that the content needs to be in the form `app@device:passwd`